janus/message-base.md

```diff
@@ -0,0 +1,306 @@
+
+Messages received with common crypto should be marked as such for internal processing. See the document about Security Context.
+### No Crypto
+There may be use cases for which crypto of any kind is difficult, such as simple IOT devices. Messages in this form send the basic payload unencoded, unsigned, and unencrypted. It is possible to include a shared secret as a bearer token.
+
+Messages received with no crypto should be marked as such for internal processing, and the actions taken intentionally limited. See the document about Security Context.
+
+## Encryption / Signing
+
+The SSI Protocol uses elliptic curve cryptography to achieve message availability,
+confidentiality, and integrity. It offers three crypto primitives:
+
+```anon_crypt(msg, recipient_verkey)```
+<blockquote>Encrypt only for recipient. Sender is anonymous. Parties can be unknown
+to each other. The message is tamper evident.</blockquote>
+
+```auth_crypt(msg, recipient_verkey, sender_sigkey)```
+<blockquote>Encrypt only for recipient. Recipient learns sender’s verkey but can’t
+prove to anybody else who the sender is (making the message repudiable).
+Parties can be unknown to each other. The message is tamper evident.</blockquote>
+
+```sign(msg, sender_sigkey)```
+<blockquote>Associate a non-repudiable signature with a message.
+Note that messages are not required to use non-repudiable signatures.</blockquote>
+
+### Signed Message
+Signing is only done when the use case requires it, for several reasons: 
+1. Authentication can be accomplished as part of the Public Key Authenticated Encryption (PKAE). 
+2. The Non-Repudiation offered by signing (DSA, ECDSA, EDDSA) is not always desired, as in the interrim steps of a negotiation. Signatures make a message non-repudiable. 
+3. Signatures are somewhat more expensive to compute and verify than other ways to authenticate, like PKAE.    
+
+If a signature is required, the original message is encoded as BASE64URL and added to a new object with two other signature fields: 
+
+```JSON
+{
+  "signed": "<string encoded message>",
+  "sig": "<string encoded signature>",
+  "sig_type": "<algorithm used to sign>"
+}
+```
+
+The `sig_type` tells us how the original message was encoded, what algorithm was used to create the signature, and how the signature is encoded.
+
+For example, the `ED25519_B64U` signature scheme has the following characteristics:
+* Pre-conditions: **None**
+* Message encoding: **BASE64URL**
+* Signing algorithm: **ed25519**
+* Signature encoding: **BASE64URL**
+
+More advanced signature schemes can be used. For example, the CL_CRED_1 algorithm has the following characteristics.
+* Pre-conditions: **message is a typed CRED version 0.1 through 0.2, JSON**
+* Message encoding: **none**
+* Signing algorithm: **CL-RSA**
+* Signature encoding: **none**
+
+### Topical Messages
+Messages are usually part of interactions.
+
+> TODO: How does this fit in with the JWT headers?
```

Comment:
Because JWT is designed for claims, and is not a general message protocol, it's "sub" (subject) claim identifies who/what the claim is about. When the message isn't a claim, but a typed message in a protocol, then the topic makes more sense. The closest thing to JWT of a  topic is the "jti" (JWT ID). Perhaps the "jti" could be co-opted for the tid, mid, and lmid, like this: 

```"jti": { "tid": "1", "mid": "1", "ptid": null, "lmid": 0}```

or in a more compact way like this:

```"jti": "1;1;;0"```